src/FSize.h

```diff
@@ -78,67 +99,87 @@ class FSize {
-      case G: return "GB";
-      case M: return "MB";
-      case K: return "kB";
-      case B: break;
+      case Unit::Y: return "YiB";
+      case Unit::Z: return "ZiB";
+      case Unit::E: return "EiB";
+      case Unit::P: return "PiB";
+      case Unit::T: return "TiB";
+      case Unit::G: return "GiB";
+      case Unit::M: return "MiB";
+      case Unit::K: return "KiB";
+      case Unit::B: break;
       }
       return "B";
     }
 
   public:
 
     /**
-     * Construct from size in Byte.
+     * Construct from size in certain unit.
+     * E.g. <code>FSize( 1, FSize::Unit::K )<code> makes 1024 Byte.
      **/
-    FSize( const long long size_r = 0 )
-      : _size( size_r )
+    FSize( const boost::multiprecision::cpp_int &size_r = 0, const Unit unit_r = Unit::B)
+      : _size( boost::multiprecision::cpp_int(size_r) * factor( unit_r ) )
     {}
 
     /**
-     * Construct from size in certain unit.
-     * E.g. <code>FSize( 1, FSize::K )<code> makes 1024 Byte.
+     * Construct from size in Byte.
+     * @param size_r the initial value
      **/
-    FSize( const long long size_r, const Unit unit_r )
-      : _size( size_r * factor( unit_r ) )
+    FSize( double size_r )
+      : _size( size_r )
     {}
 
     /**
       Construct from string containing a number in given unit.
+      * @param sizeStr input string - must contain *only* numbers
+      * @param unit_r  optional unit, bytes by default
+      * @throws std::runtime_error if the string contains any non numeric characters,
+      *   even a trailing white space!
     */
-    FSize( const std::string &sizeStr, const Unit unit_r = B );
+    FSize( const std::string &sizeStr, const Unit unit_r = Unit::B );
 
     /**
-     * Conversion to long long
+     * Conversions to native data types - only explicit as it might overflow
+     * If the value is out of range then the behavior depends on the boost version
+     *   - 1.67 - the min/max values for the corresponding type are returned
+     *   - 1.66 - returns just the lower bits
      **/
-    operator long long() const { return _size; }
+    explicit operator long long() const { return static_cast<long long>(_size); }
+    explicit operator int() const { return static_cast<int>(_size); }
+    explicit operator double() const { return static_cast<double>(_size); }
 
-    FSize & operator+=( const long long rhs ) { _size += rhs; return *this; }
-    FSize & operator-=( const long long rhs ) { _size -= rhs; return *this; }
-    FSize & operator*=( const long long rhs ) { _size *= rhs; return *this; }
-    FSize & operator/=( const long long rhs ) { _size /= rhs; return *this; }
+    operator boost::multiprecision::cpp_int() const { return _size; }
+    boost::multiprecision::cpp_int in_unit(const Unit unit_r) const { return _size / factor( unit_r ); }
 
-    FSize & operator++(/*prefix*/) { _size += 1; return *this; }
-    FSize & operator--(/*prefix*/) { _size -= 1; return *this; }
+    // unary minus
```

Comment:
Ugh, a negative size is a weird thing.
Is this for compatibility with the old FSize?
Apparently Boost does not have an unsigned multiprecision type.

Reply:
Yes, it's for compatibility.

And actually it makes sense in some cases. Libzypp reports the current disk usage and the estimated usage after installing the selected packages. If you select too many packages and there is low free space then it might happen that the packages do not fit. In that case libzypp reports required space bigger than the actual disk size.

If you want to calculate how much free space you have after installing the packages then you might get a negative number (indicating that there is not enough space).

Yes, for absolute file/disk sizes it does not make sense, but for relative numbers (a diff) it makes sense to have negative numbers.